### DISCUSSION TITLE
How to add Nextjs settings to AdminUI's build, to limit number of threads ?

### DISCUSSION THREAD

**Discussion by Francoois** [keystonejs/keystone#8291] [Questions] [answered]
Hello,
I am facing an issue that I assume is environment specific, and I think it could be avoided with appropriate new settings. So maybe it should be a new feature request.

# Environment : 
My client's server uses CloudLinux with CPanel. It's an O2Switch server, it limits the number of concurrent processes to 250.

# How to reproduce : 
On this environment, run `keystone-create-app`, then `install` and `build`.
It stops after : 
```txt
✨ Building Admin UI
info  - Skipping validation of types
info  - Skipping linting
```
Last two lines are from Next build.
At the same time many of these processes are created :
`/keystone-app/node_modules/next/dist/compiled/jest-worker/processChild.js`
The build step never complete and my CPanel monitoring tool shows `NPROC` close or equal to 250, the max.

# How the problem is addressed on vanilla NextJs :
To reproduce with Next only I installed demo app with : 
`npx create-next-app@latest nextjs-blog --use-npm --example "https://github.com/vercel/next-learn/tree/master/basics/learn-starter"`
If I run `build` : similar issue, the app stops at *Linting and checking validity of types* with same processes :
`nextjs-blog/node_modules/next/dist/compiled/jest-worker/processChild.js`

After some research turns out I can specify [workerThreads & cpu experimental](https://stackoverflow.com/questions/73938031/error-jest-worker-just-run-in-cpanel-adonisjs-nextjs-reactjs) options :
```js
module.exports = {
  experimental: {
    workerThreads: false,
    cpus: 4
  }
};
```
And it works, with these settings the Next app builds normally.

So I think it is the way to go with Keystone, but I haven't found a way to give these parameter to my AdminUI generation.

I tried to write it to `.keystone/admin/next.config.js` but it is overwritten during build.
Is there any way to proceed or should I submit a feature request ? Should I generate AdminUi separately with configured NextJS ? 
I could try to submit a PR for this feature but since I'm not familiar with KS's internals, if there is a way to do it or a workaround I would take it :)

**Comment by borisno2** [comment, marked as answer]
Hey @Francoois 
You can override the generated `next.config.js` using `ui.getAdditionalFiles` (https://keystonejs.com/docs/config/config#ui). I have an example of it here https://github.com/borisno2/keystone-in-next-auth-demo/blob/main/keystone.ts#L19 where I am copying on a new `next.config.js`- which you can see here https://github.com/borisno2/keystone-in-next-auth-demo/blob/main/next-config.js - to override the `basePath` (Disclaimer: editing the Keystone generated `next.config.js` could break the Keystone Admin UI or lead to unexpected errors)
Hope this helps!

**Reply by Francoois** [reply]
Thank you @borisno2 for the answer ! This look like what I was looking for ! :)

Note : I ended up switching to an other service provider because I had other issues, (Phusion Passenger not detecting the server's call to listen() ). If anyone managed to get it to work on O2Switch's or similar Cloudlinux/CPanel provider I'd happily take your feedbacks
